### PR TITLE
Parallel processing support

### LINKED ISSUE #241: Multithread support
### Description

Currently the registry only supports running a single instance, running a single thread to verify transactions. 

```[tasklist]
### Tasks
- [x] Add functionality to ensure all transactions for a single stream are handled syncroniosly by a single verifier thread.
```

### PR DESCRIPTION
**Pull request by MartinSchmidt**
Using RabbitMq and Consistent Hash ring algorithm, the registry now guarantees that all transactions for a certificate are sequentially handled.
It supports scaling the number of containers and threads per container up and down.
The implementation also includes a QueueCleanupService which continually monitors the RabbitMq cluster if the number of queues in the ring are lowered, then the QueueCleanupService will flush and resort the items in the inactive queues before deleting the queues.
If queues are added to the ring, items will be rerouted to the correct queue as then are processed by the TransactionProcessor if they are no longer the correct queue.

To configure this a new section to the values file was added:
```yaml
 # transactionProcessor holds the configuration for the transaction processor
 transactionProcessor:
   # replicas defines the number of transaction processor containers to run
   replicas: 3
   # threads defines the number of parallel threads to run per transaction processor
   threads: 5
   # defines the weight in the consistent hash ring used to determine which transaction processor to send a transaction to.
   weight: 10
```

fixes #241

### CODE REVIEW

File: charts/project-origin-registry/run_kind_test.sh [hunk 1]
```diff
@@ -90,10 +93,10 @@ cat << EOF > "${wallet_filename}"
 service:
   type: NodePort
   nodePort: ${wallet_nodeport}
-wallet:
+config:
   externalUrl: http://localhost:${wallet_nodeport}
-redis:
-  enabled: true
+  jwt:
+    allowAnyJwtToken: true
```
Comment (sahma19) [inline charts/project-origin-registry/run_kind_test.sh:99]: Should I be worried about this?

File: charts/project-origin-registry/run_kind_test.sh [hunk 2]
```diff
@@ -104,14 +107,14 @@ registries:
...
 echo "Wallet installed"
 
 # install two registries
 echo "Installing registries"
 helm install ${registry_a_name} -n ${registry_a_namespace} charts/project-origin-registry --set persistance.cloudNativePG.enabled=true,image.tag=test,service.nodePort=$registry_a_nodeport,service.type=NodePort -f "${override_values_filename}"  --create-namespace --wait
 echo "Registry A installed"
-helm install ${registry_b_name}-postfix -n ${registry_b_namespace} charts/project-origin-registry --set persistance.inMemory.enabled=true,image.tag=test,registryName=$registry_b_name,service.nodePort=$registry_b_nodeport,service.type=NodePort -f "${override_values_filename}"  --create-namespace --wait
+helm install ${registry_b_name}-postfix -n ${registry_b_namespace} charts/project-origin-registry --set transactionProcessor.replicas=1,persistance.inMemory.enabled=true,image.tag=test,registryName=$registry_b_name,service.nodePort=$registry_b_nodeport,service.type=NodePort -f "${override_values_filename}"  --create-namespace --wait
```
Comment (sahma19) [inline charts/project-origin-registry/run_kind_test.sh:117]: I see persistance.inMemory.enabled=true. is this intentional?

File: charts/project-origin-registry/templates/registry-deployment.yaml [hunk 3]
```diff
@@ -31,10 +31,27 @@ spec:
...
+            - name: TransactionProcessor__ServerNumber
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.labels['apps.kubernetes.io/pod-index']
+            - name: TransactionProcessor__Servers
+              value: {{ .Values.transactionProcessor.replicas | quote }}
+            - name: TransactionProcessor__Threads
+              value: {{ .Values.transactionProcessor.threads | quote }}
+            - name: TransactionProcessor__Weight
+              value: {{ .Values.transactionProcessor.weight | quote }}
```
Comment (sahma19) [inline charts/project-origin-registry/templates/registry-deployment.yaml:48]: How are the values calculated, or determined? Manually?

File: charts/project-origin-registry/templates/registry-deployment.yaml [hunk 4]
```diff
@@ -57,6 +74,46 @@ spec:
...
+            - name: RabbitMq__Username
+              valueFrom:
+                secretKeyRef:
+                  name: {{ .Release.Name }}-rabbitmq-default-user
+                  key: username
+            - name: RabbitMq__Password
+              valueFrom:
+                secretKeyRef:
+                  name: {{ .Release.Name }}-rabbitmq-default-user
+                  key: password
```
Comment (sahma19) [inline charts/project-origin-registry/templates/registry-deployment.yaml:100]: I see you are using a "rabbitmq-default-user", If a default user is used, is the default password used as well?

File: charts/project-origin-registry/values.yaml [hunk 5]
```diff
@@ -10,6 +10,30 @@ image:
...
   tag:
 
+# transactionProcessor holds the configuration for the transaction processor
+transactionProcessor:
+  # replicas defines the number of transaction processor containers to run
+  replicas: 3
+  # threads defines the number of parallel threads to run per transaction processor
+  threads: 5
+  # defines the weight in the consistent hash ring used to determine which transaction processor to send a transaction to.
+  weight: 10
```
Comment (sahma19) [inline charts/project-origin-registry/values.yaml:20]: I believe this answers my other question, about how they are calculated. Manually it seems. Should it occur, that the throughput exceeds what can be handled, by the replica, I assume another replica will be spun up? 

Have you made any thoughts about whether:

More threads, per replica vs. a new replica, for handling high throughput?

File: charts/project-origin-registry/values.yaml [hunk 6]
```diff
@@ -95,6 +118,11 @@ immutableRecord:
       # key is the reference to the value within the secret
       key: ""
 
+# redis holds the configuration for the redis subchart
+redis:
+  # enabled specifies if the redis subchart should be enabled or not
+  enabled: true
+
```
Comment (sahma19) [inline charts/project-origin-registry/values.yaml:125]: What would happen if the subchart was disabled? would Redis stop being deployed?

File: src/ProjectOrigin.Electricity.Example/WithoutWalletFlow.cs [hunk 7]
```diff
@@ -76,7 +76,7 @@ public async Task<int> Run()
 
 
         // ------------------  allocate ------------------
-        Console.WriteLine($"Allocating Production Granular Certificate");
+        Console.WriteLine($"Creating allocation event");
```
Comment (sahma19) [inline src/ProjectOrigin.Electricity.Example/WithoutWalletFlow.cs:79]: Why this change?

File: src/ProjectOrigin.Registry.IntegrationTests/ConsistentHashResolverTests.cs [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+            Servers = 2,
+            Threads = 2,
+            Weight = 2
+        };
+        var resolver = new ConsistentHashRingQueueResolver(Options.Create(options));
+
+        var queue = resolver.GetQueueName("test");
+
+        Assert.Equal("registry_0.verifier_1", queue);
+    }
```
Comment (sahma19) [inline src/ProjectOrigin.Registry.IntegrationTests/ConsistentHashResolverTests.cs:59]: I see that you are not using the same amount of threads, weight, and replicas for tests as in prod. is this integration test then valid enough?

File: src/ProjectOrigin.Registry.Server/Services/RabbitMqHttpClient.cs [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+using ProjectOrigin.Registry.Server.Interfaces;
+using ProjectOrigin.Registry.Server.Models;
+using ProjectOrigin.Registry.Server.Options;
+
+namespace ProjectOrigin.Registry.Server.Services;
+
+public sealed class RabbitMqHttpClient : IRabbitMqHttpClient, IDisposable
+{
+    private const string Protocol = "http";
+    private readonly HttpClient _httpClient;
```
Comment (ckr123) [inline src/ProjectOrigin.Registry.Server/Services/RabbitMqHttpClient.cs:18]: Would it make sense to use the HttpClientFactory here instead? 

Having a hard time figuring out if there is a captive dependency here

File: src/ProjectOrigin.Registry.IntegrationTests/Extensions/RandomExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+    public static class RandomExtensions
+    {
+        public static string GenerateString(this Random r, int length = 10)
+        {
+            var bytes = new byte[length];
+            r.NextBytes(bytes);
+            return Convert.ToBase64String(bytes);
+        }
+    }
+}
```
Comment (sahma19) [inline src/ProjectOrigin.Registry.IntegrationTests/Extensions/RandomExtensions.cs:14]: Is this to simulate WalletDepositEndpoints?

File: src/ProjectOrigin.Registry.IntegrationTests/PerformanceTests.cs [hunk 11]
```diff
@@ -72,6 +77,15 @@ public class PerformanceTests : IAsyncLifetime, IClassFixture<ContainerImageFixt
                 .WithEnvironment("Logging__LogLevel__Grpc.AspNetCore", "Information")
                 .WithEnvironment("Cache__Type", "redis")
                 .WithEnvironment("Cache__Redis__ConnectionString", redisFixture.ContainerConnectionString)
+                .WithEnvironment("RabbitMq__Hostname", rabbitMqFixture.ContainerIp)
+                .WithEnvironment("RabbitMq__AmqpPort", RabbitMqFixture.ContainerAmqpPort.ToString())
+                .WithEnvironment("RabbitMq__HttpApiPort", RabbitMqFixture.ContainerHttpPort.ToString())
```
Comment (sahma19) [inline src/ProjectOrigin.Registry.IntegrationTests/PerformanceTests.cs:82]: The port is dynamic correct?

File: src/ProjectOrigin.Registry.IntegrationTests/PerformanceTests.cs [hunk 12]
```diff
@@ -137,7 +151,7 @@ public async Task TestThroughput()
         Console.WriteLine($"Completed {completed.Count} transactions in {elapsedSeconds} seconds ({requestsPerSecond} requests per second).");
         Console.WriteLine($"-- Finished throughput test --");
 
-        requestsPerSecond.Should().BeGreaterThan(30); // based on througput test on github ~35
+        requestsPerSecond.Should().BeGreaterThan(150); // based on througput test on github ~170
```
Comment (sahma19) [inline src/ProjectOrigin.Registry.IntegrationTests/PerformanceTests.cs:154]: What exactly is tested here? could a situation ever occur where sent requests =/= requestsPerSecond?

**Review by sahma19 — CHANGES_REQUESTED** [review]

File: charts/project-origin-registry/run_kind_test.sh (hunk 2 above)
Comment (MartinSchmidt) [reply charts/project-origin-registry/run_kind_test.sh:117]: Yes, part of a full test where two registries are spun up, one inMemory, and one using postgresql to test both and how they interop. The one using inMemory, can only run a single replica for good reason, otherwise data would be inconsistent between pods.

File: charts/project-origin-registry/run_kind_test.sh (hunk 1 above)
Comment (MartinSchmidt) [reply charts/project-origin-registry/run_kind_test.sh:99]: This is only in the test, as not to have to mess with it. 
This is a test of the registries, and not the wallet, just using the wallet system for the test.

File: charts/project-origin-registry/templates/registry-deployment.yaml (hunk 3 above)
Comment (MartinSchmidt) [reply charts/project-origin-registry/templates/registry-deployment.yaml:48]: They come from the values.yaml file, they are determined by the ones deploying the registry.

File: charts/project-origin-registry/templates/registry-deployment.yaml (hunk 4 above)
Comment (MartinSchmidt) [reply charts/project-origin-registry/templates/registry-deployment.yaml:100]: the default-user that is referred to here are generated by the [rabbitmq operator and stored in a secret](https://www.rabbitmq.com/kubernetes/operator/using-operator.html#creds) so they are unique and not some default values people can find online.

File: charts/project-origin-registry/values.yaml (hunk 5 above)
Comment (MartinSchmidt) [reply charts/project-origin-registry/values.yaml:20]: Currently this is done manually, automatic throttling would be a separate issue to implement.
Pros and cons, the number of queues in the hash-ring is replicas * threads. 
Replicas ensure one can do a rolling update and split the workload or more physical machines in a cluster. Threads is cheaper since an extra contrainer is not spun up.

File: charts/project-origin-registry/values.yaml (hunk 6 above)
Comment (MartinSchmidt) [reply charts/project-origin-registry/values.yaml:125]: yes, they they fallback to a inMemory cache per node, it is not recommended when running multiple nodes.

File: src/ProjectOrigin.Electricity.Example/WithoutWalletFlow.cs (hunk 7 above)
Comment (MartinSchmidt) [reply src/ProjectOrigin.Electricity.Example/WithoutWalletFlow.cs:79]: originally a typo, same log statement few lines later..

File: src/ProjectOrigin.Registry.IntegrationTests/ConsistentHashResolverTests.cs (hunk 8 above)
Comment (MartinSchmidt) [reply src/ProjectOrigin.Registry.IntegrationTests/ConsistentHashResolverTests.cs:59]: this is a unit-test of the ConsistentHashRingQueueResolver and its functionality is tested in isolation.
(yes i know it is in a project named IntegrationTests, currently all tests are in the same project.)

File: src/ProjectOrigin.Registry.IntegrationTests/Extensions/RandomExtensions.cs (hunk 10 above)
Comment (MartinSchmidt) [reply src/ProjectOrigin.Registry.IntegrationTests/Extensions/RandomExtensions.cs:14]: just a shorthand to easily generate base64 strings of certain byte lengths.

File: src/ProjectOrigin.Registry.IntegrationTests/PerformanceTests.cs (hunk 11 above)
Comment (MartinSchmidt) [reply src/ProjectOrigin.Registry.IntegrationTests/PerformanceTests.cs:82]: could be redirected in kubernetes services, so therefore it is possibly to configure them.

File: src/ProjectOrigin.Registry.Server/Services/RabbitMqHttpClient.cs (hunk 9 above)
Comment (MartinSchmidt) [reply src/ProjectOrigin.Registry.Server/Services/RabbitMqHttpClient.cs:18]: Done :)

Comment (ckr123) [reply src/ProjectOrigin.Registry.Server/Services/RabbitMqHttpClient.cs:18]: Was actual thinking of https://learn.microsoft.com/en-us/dotnet/core/extensions/httpclient-factory because it looks to me like this client is getting the same lifetime as the background service implementing the RabbitMqHttpClient and i believe this should be a short lived httpclient

https://blog.ploeh.dk/2014/06/02/captive-dependency/
https://learn.microsoft.com/en-us/dotnet/fundamentals/networking/http/httpclient-guidelines

**Review by sahma19 — APPROVED** [review]
I have read your answers, to my comments. Approving.
